src/main/java/com/spikes2212/path/PurePursuitController.java

```diff
@@ -0,0 +1,128 @@
+package com.spikes2212.path;
+
+/**
+ * This class represents a PurePursuitController.
+ * You should the getTargetSpeed methods periodically.
+ *
+ * @author T
+ */
+public class PurePursuitController {
+    private OdometryHandler handler;
+    private Path path;
+    private int lastClosestIndex = 0, lastLookaheadIndex = 0;
+    private double lookaheadDistance;
+    private double robotWidth;
+
+    public PurePursuitController(OdometryHandler handler, Path path, double lookaheadDistance,
+                                 double robotWidth) {
+        this.handler = handler;
+        this.path = path;
+        this.lookaheadDistance = lookaheadDistance;
+        this.robotWidth = robotWidth;
+    }
+
+    public OdometryHandler getHandler() {
+        return handler;
+    }
+
+    public void setHandler(OdometryHandler handler) {
+        this.handler = handler;
+    }
+
+    public Path getPath() {
+        return path;
+    }
+
+    public void setPath(Path path) {
+        this.path = path;
+    }
+
+    public double getLookaheadDistance() {
+        return lookaheadDistance;
+    }
+
+    public void setLookaheadDistance(double lookaheadDistance) {
+        this.lookaheadDistance = lookaheadDistance;
+    }
+
+    private Waypoint closestPoint() {
+        Waypoint robot = handler.getWaypoint();
+        double minDistance = Double.POSITIVE_INFINITY, distance;
+        int minIndex = lastClosestIndex;
+        for (int i = lastClosestIndex; i < path.getPoints().size(); i++) {
+            if ((distance = path.getPoints().get(i).distance(robot)) < minDistance) {
+                minIndex = i;
+                minDistance = distance;
+            }
+        }
+        lastClosestIndex = minIndex;
+        return path.getPoints().get(minIndex);
+    }
+
+    private Waypoint getLookaheadPoint() throws LookaheadPointNotFoundException {
```

Comment:
please add a `lastLookaheadPoint` variable which will be updated each time `getLookaheadPoint` is called. and then it is returned from getLookaheadPoint. This is needed in order to make the point stay at the end of the path, when the robot starts to approach it's end (when the distance to the lookahead point is below the given distance).